reference/src/representation/pointers.md

```diff
@@ -0,0 +1,43 @@
+# Representation of reference and pointer types
+
+### Terminology
+
+Reference types are types of the form `&T` or `&mut T`.
+
+Raw pointer types are types of the form `*const T` or `*mut T`.
+
+### Representation
+
+The alignment of reference and raw pointer types is the word size.
```

Comment:
I just realized this might not actually be true for custom DSTs (e.g., consider a custom DST with `u64` metadata on a 32 bit target). Since the rest of this text is already careful to be forward compatible with custom DSTs, it would be a shame if this part wasn't.

Reply:
Hmm, worth thinking about. On 32-bit targets, does `u64` have 64-bit alignment? Are there architectures with alignment larger than the word size?

Reply:
Oh, u64 is not a good example, because on most (all?) 32 bit targets it has alignment 4. But there are definitely tons of types with super-word alignment -- anyone can define them with `#[repr(align(N))]`, and vector types such as `__m128` usually require natural alignment (i.e., size = align).

Reply:
So the question is do we want to tie our hands and say that metadata of custom DSTs is at most word-aligned? The trade-offs don't seem obvious to me, I'm trying to come up with an example and the best I can do is something like:
```rust
  let a: Box<T> = Box::new(T::new());
  let b: Box<(*T, usize, usize)> = Box::new(Box::Into_raw(a), 0, 0);
  let c: &&U = unsafe { mem::transmute(Box::into_raw(b)) };
```
The question is can this be insta-UB if `U`s metadata is super-word aligned?

This is the simplest example I can think of.

Reply:
`u64` has 8-byte alignment on Windows and MacOS 32-bit.

Reply:
My two cents is that we shouldn't say anything about the size and alignment of references here. 

We should just say that the layout of references and pointers is the same as the layout of the following `repr(Rust)` type:

```rust
struct Layout<Metadata> {
    ptr: *mut (), 
    metadata: Metadata
}
```

We can then say that in `*mut T`/`&mut T`, if `T` is a trait / slice / ... then `Metadata`'s layout is the same as the layout of some other particular type, and the layout of references and pointers follows from that.

Reply:
When we have custom DSTs in the language, then *all* pointers are `... { ptr: *mut (), metadata: T::Metadata }`, as custom DSTs generalize slices, trait objects, and sized types (we can still spell this out for clarity but there's no need to have two things called `Metadata` or to introduce this `===` relation). But we don't have custom DSTs yet, so it feels very weird to define layout in terms of a concept the language doesn't have yet, even if we can technically avoid any reference to a specific non-existent `DynamicallySized` trait.

> Since I've used raw pointers, we have to say something about the size and alignment of raw pointers here, but we already said that their size is the same as usize, and for alignment we can just say that the raw pointer alignment is an implementation-defined multiple of their size.

Wait, what? If the alignment is not described yet, it should be described if we can at all get consensus for it. As far as I know, that alignment should be *exactly* the size. But even if that is controversial, it's the other way around, size has to be a multiple of the alignment (because element i of an array is at byte offset `i * alignof(T)`).

Reply:
>  But we don't have custom DSTs yet, so it feels very weird to define layout in terms of a concept the language doesn't have yet, even if we can technically avoid any reference to a specific non-existent DynamicallySized trait.

Even without the concept of custom DSTs, we do have the concept of generic structs, and their layout, and those can be used to specify the layout of slices and trait objects.

Reply:
Sure, but unifying these two layouts into one generic struct, and the naming of the type parameter, gestures heavily at the concept of custom DSTs (would anyone make this suggestion if they didn't know of the generalization that custom DSTs bring?), without really buying us anything right now.

Reply:
OK, I reworded this to say that reference types have at least word alignment, and have exactly word alignment in the case of references to sized types, `&[T]`, `&str` or `&dyn Trait`.